File: MyScript/edk/c/examples/shape/MultipleShapesRecognitionBeautification.c
/**
 * The `MultipleShapesRecognitionBeautification` example code shows you how to
 * perform recognition and beautification of multiple handwritten shapes.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptShape.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// digital ink
#include "rectangle.strokes.h"
#include "arrow.strokes.h"
#include "circle.strokes.h"
#include "triangle.strokes.h"

// forward declaration of example functions
static void printResult(voShapeDocument document);

// main function
int main()
{
  voShapeRecognizer recognizer = NULL;
  voShapeBeautifier beautifier = NULL;
  voShapeKnowledge shapeKnowledge = NULL;
  voShapeDocument document = NULL;

  // -- initialization ---------------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) creating a new ShapeRecognizer and a new ShapeBeautifier\n");
  recognizer = voCreateShapeRecognizer(engine);
  verify(recognizer != NULL);
  printf(" . ShapeRecognizer created successfully\n");
  beautifier = voCreateShapeBeautifier(engine);
  verify(beautifier != NULL);
  printf(" . ShapeBeautifier created successfully\n");
  printf("\n");

  printf("3) loading and attaching the ShapeKnowledge\n");
  shapeKnowledge = loadEngineObject(SHAPE_KNOWLEDGE_RES);
  printf(" . ShapeKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, shapeKnowledge));
  printf(" . ShapeKnowledge resource successfully attached to the ShapeRecognizer\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, beautifier, shapeKnowledge));
  printf(" . ShapeKnowledge resource successfully attached to the ShapeBeautifier\n");
  printf(" . releasing the ShapeKnowledge resource (not needed anymore)\n");
  verify(voReleaseObject(engine, shapeKnowledge));
  printf(" . ShapeKnowledge resource released successfully\n");
  printf("\n");

  printf("4) creating a ShapeDocument object\n");
  document = voCreateShapeDocument(engine);
  verify(document != NULL);
  printf(" . ShapeDocument created successfully\n");
  printf("\n");

  printf("5) fueling the ShapeDocument with digital ink\n");
  printf(" . adding a rectangle\n");
  printf("   ,--------.\n");
  printf("   |        |\n");
  printf("   |        |\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s0[0], 2 * sizeof(*rectangle_s0), &rectangle_s0[1], 2 * sizeof(*rectangle_s0), (uint32_t)(sizeof(rectangle_s0) / sizeof(*rectangle_s0)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s1[0], 2 * sizeof(*rectangle_s1), &rectangle_s1[1], 2 * sizeof(*rectangle_s1), (uint32_t)(sizeof(rectangle_s1) / sizeof(*rectangle_s1)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s2[0], 2 * sizeof(*rectangle_s2), &rectangle_s2[1], 2 * sizeof(*rectangle_s2), (uint32_t)(sizeof(rectangle_s2) / sizeof(*rectangle_s2)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &rectangle_s3[0], 2 * sizeof(*rectangle_s3), &rectangle_s3[1], 2 * sizeof(*rectangle_s3), (uint32_t)(sizeof(rectangle_s3) / sizeof(*rectangle_s3)) / 2));

  printf(" . adding an arrow\n");
  printf("   ,--------.\n");
  printf("   |        |\n");
  printf("   |        | ------>\n");
  printf("   '--------'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &arrow_s0[0], 2 * sizeof(*arrow_s0), &arrow_s0[1], 2 * sizeof(*arrow_s0), (uint32_t)(sizeof(arrow_s0) / sizeof(*arrow_s0)) / 2));
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &arrow_s1[0], 2 * sizeof(*arrow_s1), &arrow_s1[1], 2 * sizeof(*arrow_s1), (uint32_t)(sizeof(arrow_s1) / sizeof(*arrow_s1)) / 2));

  printf(" . adding a circle\n");
  printf("                        ,---.\n");
  printf("   ,--------.          /     \\\n");
  printf("   |        |         |       |\n");
  printf("   |        | ------>  \\     /\n");
  printf("   '--------'           '---'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &circle_s0[0], 2 * sizeof(*circle_s0), &circle_s0[1], 2 * sizeof(*circle_s0), (uint32_t)(sizeof(circle_s0) / sizeof(*circle_s0)) / 2));

  printf(" . adding a triangle\n");
  printf("                        ,---.\n");
  printf("   ,--------.          /     \\\n");
  printf("   |        |         |       |\n");
  printf("   |        | ------>  \\     /\n");
  printf("   '--------'           '---'\n");
  printf("                  .\n");
  printf("                 / \\\n");
  printf("                /   \\\n");
  printf("               '-----'\n");
  verify(VO_IFACE(engine, IShapeDocument)->addStroke(engine, document, &triangle_s0[0], 2 * sizeof(*triangle_s0), &triangle_s0[1], 2 * sizeof(*triangle_s0), (uint32_t)(sizeof(triangle_s0) / sizeof(*triangle_s0)) / 2));
  printf(" . ShapeDocument successfully fueled with digital ink\n");
  printf("\n");

  printf("6) running the shape recognition process\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, recognizer, document, NULL, NULL));
  printf(" . shape recognition performed successfully\n");
  printf("\n");

  printf("7) retrieving the recognition result\n");
  printResult(document);
  printf("\n");

  printf("8) running the shape beautification process\n");
  verify(VO_IFACE(engine, IShapeDocumentProcessor)->process(engine, beautifier, document, NULL, NULL));
  printf(" . shape beautification performed successfully\n");
  printf("\n");

  printf("9) retrieving the beautification result\n");
  printResult(document);
  printf("\n");

  printf("10) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, document));
  verify(voReleaseObject(engine, recognizer));
  verify(voReleaseObject(engine, beautifier));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform multiple shapes recognition and beautification\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}

static void printResult(voShapeDocument document)
{
  uint32_t segmentCount = 0;
  uint32_t i = 0;

  printf(" . retrieving the number of ShapeSegments in the ShapeDocument: ");
  segmentCount = VO_IFACE(engine, IShapeDocument)->getSegmentCount(engine, document);
  verify(segmentCount != (uint32_t)-1);
  printf("%d\n", segmentCount);

  printf(" . iterating over the ShapeSegments:\n");
  for (i = 0; i < segmentCount; ++i)
  {
    voShapeSegment segment = NULL;
    uint32_t candidateCount = 0;

    segment = VO_IFACE(engine, IShapeDocument)->getSegmentAt(engine, document, i);
    verify(segment != NULL);

    candidateCount = VO_IFACE(engine, IShapeSegment)->getCandidateCount(engine, segment);
    verify(candidateCount != (uint32_t)-1);

    if (candidateCount != 0)
    {
      voShapeCandidate candidate = NULL;

      // get the best candidate
      candidate = VO_IFACE(engine, IShapeSegment)->getCandidateAt(engine, segment, 0);
      verify(candidate != NULL);

      switch(voGetType(engine, candidate))
      {
        case VO_ShapeRecognized:
          {
            voShapeModel model = NULL;
            voString label = {NULL, 0};
            uint32_t primitiveCount = 0;
            uint32_t j = 0;

            float rs = 0;
            float nrs = 0;

            model = VO_IFACE(engine, IShapeRecognized)->getModel(engine, candidate);
            verify(model != NULL);

            verify(VO_IFACE(engine, IShapeModel)->getLabel(engine, model, NULL, &label));
            label.bytes = malloc(label.byteCount + 1);
            verify(VO_IFACE(engine, IShapeModel)->getLabel(engine, model, NULL, &label));
            label.bytes[label.byteCount] = '\0';

            rs = VO_IFACE(engine, IShapeRecognized)->getResemblanceScore(engine, candidate);
            nrs = VO_IFACE(engine, IShapeRecognized)->getNormalizedRecognitionScore(engine, candidate);

            printf("   . segment %d, best candidate: %s (rs:%.4f - nrs:%.4f)\n", i, label.bytes, rs, nrs);

            primitiveCount = VO_IFACE(engine, IShapeRecognized)->getPrimitiveCount(engine, candidate);
            verify(primitiveCount != (uint32_t)-1);

            for (j = 0; j < primitiveCount; ++j)
            {
              voShapePrimitive primitive = NULL;

              primitive = VO_IFACE(engine, IShapeRecognized)->getPrimitiveAt(engine, candidate, j);
              verify(primitive != NULL);

              switch(voGetType(engine, primitive))
              {
                case VO_ShapeLine:
                  {
                    voShapeLineData data;

                    verify(VO_IFACE(engine, IShapeLine)->getData(engine, primitive, &data));
                    printf( "     . line: (%.2f, %.2f), (%.2f, %.2f)\n",
                            data.p1.x, data.p1.y,
                            data.p2.x, data.p2.y);
                  }
                  break;

                case VO_ShapeEllipticArc:
                  {
                    voShapeEllipticArcData data;

                    verify(VO_IFACE(engine, IShapeEllipticArc)->getData(engine, primitive, &data));
                    printf( "     . arc: (%.2f, %.2f), %.2f, %.2f, %.2f, %.2f, %.2f\n",
                            data.center.x, data.center.y,
                            data.minRadius,
                            data.maxRadius,
                            data.orientation,
                            data.startAngle,
                            data.sweepAngle);
                  }
                  break;

                case VO_ShapeDecoratedLine:
                  {
                    voShapeDecoratedLineData data;

                    verify(VO_IFACE(engine, IShapeDecoratedLine)->getData(engine, primitive, &data));
                    printf( "     . decorated line: (%.2f, %.2f)%s, (%.2f, %.2f)%s, \n",
                            data.line.p1.x, data.line.p1.y,
                            data.p1Decoration == VO_SHAPE_DECORATION_ARROW_HEAD ? " head" : "",
                            data.line.p2.x, data.line.p2.y,
                            data.p2Decoration == VO_SHAPE_DECORATION_ARROW_HEAD ? " head" : "");
                  }
                  break;

                default:
                  break;
              }

              verify(voReleaseObject(engine, primitive));
            }

            free(label.bytes);
            // don't forget to release objects returned by the API once you're done
            verify(voReleaseObject(engine, model));
          }
          break;

        case -1:
          printf("unable to get candidate type\n");
          break;

        default:
          // ignore other types of candidates
          break;
      }

      verify(voReleaseObject(engine, candidate));
    }

    verify(voReleaseObject(engine, segment));
  }
}

File: MyScript/edk/java/examples/shape/MultipleShapesRecognitionBeautification.java
package shape;

import java.text.NumberFormat;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.shape.DecorationType;
import com.myscript.shape.ShapeBeautifier;
import com.myscript.shape.ShapeCandidate;
import com.myscript.shape.ShapeDecoratedLine;
import com.myscript.shape.ShapeDecoratedLineData;
import com.myscript.shape.ShapeDocument;
import com.myscript.shape.ShapeEllipticArc;
import com.myscript.shape.ShapeEllipticArcData;
import com.myscript.shape.ShapeKnowledge;
import com.myscript.shape.ShapeLine;
import com.myscript.shape.ShapeLineData;
import com.myscript.shape.ShapeModel;
import com.myscript.shape.ShapePointData;
import com.myscript.shape.ShapePrimitive;
import com.myscript.shape.ShapeRecognized;
import com.myscript.shape.ShapeRecognizer;
import com.myscript.shape.ShapeSegment;
import common.CFG;

/**
 * The <code>MultipleShapesRecognitionBeautification</code> example code shows
 * you how to perform recognition and beautification of multiple handwritten
 * shapes.
 */
public class MultipleShapesRecognitionBeautification
{
  /**
   * Main entry point for the
   * <code>MultipleShapesRecognitionBeautification</code> example program.
   *
   * @param args command line arguments, ignored.
   */
  public static final void main(String[] args)
  {
    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) creating a new ShapeRecognizer and a new ShapeBeautifier");
    final ShapeRecognizer recognizer = ShapeRecognizer.create(engine);
    System.out.println(" . ShapeRecognizer created successfully");
    final ShapeBeautifier beautifier = ShapeBeautifier.create(engine);
    System.out.println(" . ShapeBeautifier created successfully");
    System.out.println();

    System.out.println("3) loading and attaching the ShapeKnowledge to the ShapeRecognizer");
    final ShapeKnowledge shapeKnowledge = (ShapeKnowledge)EngineObject.load(engine, CFG.getShapeKnowledgeRes());
    System.out.println(" . ShapeKnowledge resource loaded successfully");
    recognizer.attach(shapeKnowledge);
    System.out.println(" . ShapeKnowledge resource attached successfully to the ShapeRecognizer");
    beautifier.attach(shapeKnowledge);
    System.out.println(" . ShapeKnowledge resource attached successfully to the ShapeBeautifier");
    System.out.println(" . disposing of the ShapeKnowledge resource (not needed anymore)");
    shapeKnowledge.dispose();
    System.out.println(" . ShapeKnowledge resource disposed of successfully");
    System.out.println();
    
    System.out.println("4) creating a new ShapeDocument");
    final ShapeDocument document = ShapeDocument.create(engine);
    System.out.println(" . ShapeDocument created successfully");
    System.out.println();

    System.out.println("5) fueling the ShapeDocument with digital ink");
    System.out.println(" . adding a rectangle");
    System.out.println("   ,--------.");
    System.out.println("   |        |");
    System.out.println("   |        |");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.RECTANGLE.length; ++i)
    {
      float[] xy = Ink.RECTANGLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding an arrow");
    System.out.println("   ,--------.");
    System.out.println("   |        |");
    System.out.println("   |        | ------>");
    System.out.println("   '--------'");
    for (int i = 0; i < Ink.ARROW.length; ++i)
    {
      float[] xy = Ink.ARROW[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding a circle");
    System.out.println("                        ,---.");
    System.out.println("   ,--------.          /     \\");
    System.out.println("   |        |         |       |");
    System.out.println("   |        | ------>  \\     /");
    System.out.println("   '--------'           '---'");
    for (int i = 0; i < Ink.CIRCLE.length; ++i)
    {
      float[] xy = Ink.CIRCLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    
    System.out.println(" . adding a triangle");
    System.out.println("                        ,---.");
    System.out.println("   ,--------.          /     \\");
    System.out.println("   |        |         |       |");
    System.out.println("   |        | ------>  \\     /");
    System.out.println("   '--------'           '---'");
    System.out.println("                  .");
    System.out.println("                 / \\");
    System.out.println("                /   \\");
    System.out.println("               '-----'");
    for (int i = 0; i < Ink.TRIANGLE.length; ++i)
    {
      float[] xy = Ink.TRIANGLE[i];
      
      document.addStroke(xy, 0, 2, xy, 1, 2, xy.length / 2);
    }
    System.out.println(" . ShapeDocument successfully fueled with digital ink");
    System.out.println();
    
    System.out.println("6) running the shape recognition process");
    recognizer.process(document);
    System.out.println(" . shape recognition performed successfully");
    System.out.println();
    
    System.out.println("7) retrieving the recognition result");
    printResult(document);
    System.out.println();

    System.out.println("8) running the shape beautification process");
    beautifier.process(document);
    System.out.println(" . shape beautification performed successfully");
    System.out.println();
    
    System.out.println("9) retrieving the beautification result");
    printResult(document);
    System.out.println();
    
    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("10) cleaning-up: disposing of objects and disposing of the Engine");
    document.dispose();
    recognizer.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();
    
    System.out.println("end of example:");
    System.out.println(" you just learnt how to perform multiple shapes recognition and beautification");
  }
  
  private static void printResult(final ShapeDocument document)
  {
    System.out.print(" . retrieving the number of ShapeSegments in the ShapeDocument: ");
    final int segmentCount = document.getSegmentCount();
    System.out.println(segmentCount);
    
    final NumberFormat nf = NumberFormat.getInstance();
    nf.setGroupingUsed(false);
    
    System.out.println(" . iterating over the ShapeSegments:");
    for (int i = 0; i < segmentCount; ++i)
    {
      final ShapeSegment segment = document.getSegmentAt(i);
      final int candidateCount = segment.getCandidateCount();

      if (candidateCount != 0)
      {
        final ShapeCandidate candidate = segment.getCandidateAt(0);

        if (candidate instanceof ShapeRecognized)
        {
          final ShapeModel model = ((ShapeRecognized)candidate).getModel();

          final float rs = ((ShapeRecognized)candidate).getResemblanceScore();
          final float nrs = ((ShapeRecognized)candidate).getNormalizedRecognitionScore();

          nf.setMinimumFractionDigits(4);
          nf.setMaximumFractionDigits(4);
          System.out.println("   . segment " + i + ", best candidate: " + model.getLabel() +" (rs:" + nf.format(rs) + " nrs:" + nf.format(nrs) + ')');
          
          final int primitiveCount = ((ShapeRecognized)candidate).getPrimitiveCount();
          
          nf.setMinimumFractionDigits(2);
          nf.setMaximumFractionDigits(2);
          for (int j = 0; j < primitiveCount; ++j)
          {
            final ShapePrimitive primitive = ((ShapeRecognized)candidate).getPrimitiveAt(j);
           
            if (primitive instanceof ShapeLine)
            {
              final ShapeLineData data = ((ShapeLine)primitive).getData();
              final ShapePointData p1 = data.getP1();
              final ShapePointData p2 = data.getP2();
              
              System.out.println( "     . line: (" + nf.format(p1.getX()) + ", " + nf.format(p1.getY()) +
                                  "), (" + nf.format(p2.getX()) + ", " + nf.format(p2.getY()) + ')');
            }
            else if (primitive instanceof ShapeEllipticArc)
            {
              final ShapeEllipticArcData data = ((ShapeEllipticArc)primitive).getData();
              final ShapePointData center = data.getCenter();
             
              System.out.println( "     . arc: (" + nf.format(center.getX()) + ", " + nf.format(center.getY()) + ')' +
                                  ", " + nf.format(data.getMinRadius()) + 
                                  ", " + nf.format(data.getMaxRadius()) +
                                  ", " + nf.format(data.getOrientation()) +
                                  ", " + nf.format(data.getStartAngle()) + 
                                  ", " + nf.format(data.getSweepAngle()));
            }
            else if (primitive instanceof ShapeDecoratedLine)
            {
              final ShapeDecoratedLineData data = ((ShapeDecoratedLine)primitive).getData();
              final ShapePointData p1 = data.getLine().getP1();
              final ShapePointData p2 = data.getLine().getP2();
              
              System.out.println( "     . decorated line: (" + nf.format(p1.getX()) + ", " + nf.format(p1.getY()) + ')' + (data.getP1Decoration().equals(DecorationType.ARROW_HEAD) ? " head" : "") +
                                  ", (" + nf.format(p2.getX()) + ", " + nf.format(p2.getY()) + ')' + (data.getP2Decoration().equals(DecorationType.ARROW_HEAD) ? " head" : ""));
              
            }
            
            primitive.dispose();
          }
          
          model.dispose();
        }

        candidate.dispose();
      }

      segment.dispose();
    }
  }

} // MultipleShapesRecognitionBeautification
